Packs/Code42/Integrations/Code42/Code42.py

```diff
@@ -15,509 +17,692 @@
+    search_args.append_result(exposure, _create_exposure_filter)
+
+    query = search_args.to_all_query()
+    LOG("File Event Query: {}".format(str(query)))
+    return query
+
+
+def _create_hash_filter(hash_arg):
+    if not hash_arg:
+        return None
+    elif len(hash_arg) == 32:
+        return MD5.eq(hash_arg)
+    elif len(hash_arg) == 64:
+        return SHA256.eq(hash_arg)
+
+
+def _create_exposure_filter(exposure_arg):
+    # Because the CLI can't accept lists, convert the args to a list if the type is string.
+    if isinstance(exposure_arg, str):
+        exposure_arg = exposure_arg.split(",")
+    return ExposureType.is_in(exposure_arg)
+
+
+def _create_category_filter(file_type):
+    category_value = CODE42_FILE_TYPE_MAPPER.get(file_type["category"], "UNCATEGORIZED")
+    return FileCategory.eq(category_value)
+
+
+class ObservationToSecurityQueryMapper(object):
+    """Class to simplify the process of mapping observation data to query objects."""
+
+    _ENDPOINT_TYPE = "FedEndpointExfiltration"
+    _CLOUD_TYPE = "FedCloudSharePermissions"
+    _PUBLIC_SEARCHABLE = "PublicSearchableShare"
+    _PUBLIC_LINK = "PublicLinkShare"
+
+    def __init__(self, observation, actor):
+        self._obs = observation
+        self._actor = actor
+
+    @property
+    def _observation_data(self):
+        return self._obs["data"]
+
+    @property
+    def _exfiltration_type(self):
+        return self._obs["type"]
+
+    @property
+    def _is_endpoint_exfiltration(self):
+        return self._exfiltration_type == self._ENDPOINT_TYPE
+
+    @property
+    def _is_cloud_exfiltration(self):
+        return self._exfiltration_type == self._CLOUD_TYPE
+
+    def _create_user_filter(self):
+        return (
+            DeviceUsername.eq(self._actor)
+            if self._is_endpoint_exfiltration
+            else Actor.eq(self._actor)
+        )
+
+    def map(self):
+        search_args = self._create_search_args()
+        query = search_args.to_all_query()
+        LOG("Alert Observation Query: {}".format(query))
+        return query
+
+    def _create_search_args(self):
+        filters = FileEventQueryFilters()
+        exposure_types = self._observation_data["exposureTypes"]
+        begin_time = _convert_date_arg_to_epoch(self._observation_data["firstActivityAt"])
+        end_time = _convert_date_arg_to_epoch(self._observation_data["lastActivityAt"])
+
+        filters.append(self._create_user_filter())
+        filters.append(EventTimestamp.on_or_after(begin_time))
+        filters.append(EventTimestamp.on_or_before(end_time))
+        filters.extend(self._create_exposure_filters(exposure_types))
+        filters.append(self._create_file_category_filters())
+
+        return filters
+
+    def _create_exposure_filters(self, exposure_types):
+        """Determine exposure types based on alert type"""
+
+        if self._is_cloud_exfiltration:
+            exp_types = []
+            if self._PUBLIC_SEARCHABLE in exposure_types:
```

Comment:
This is where we would want to handle `OutsideTrustedDomain` like Nathan brought up yesterday